docs/sql/sql.md

```diff
@@ -0,0 +1,440 @@
+# SQL Support
+
+SpacetimeDB supports two subsets of SQL.
+One for ad hoc queries against the database.
+These can be made through the [cli] or the [http] endpoint.
+Another for subscriptions made through the [websocket] api or [sdk].
+
+## Subscriptions
+
+```ebnf
+SELECT projection FROM relation [ WHERE predicate ]
+```
+
+The subscription language is a query language.
+It does not support data manipulation commands.
+It is used to subscribe to a set of rows from a single table.
+
+### SELECT
+
+```ebnf
+SELECT ( '*' | table '.' '*' )
+```
+
+The `SELECT` clause determines the table that is being subscribed to.
+A wildcard projection `*` is allowed when the table is unambiguous.
+A qualified wildcard projection `.*` is needed when there are multiple tables.
+Column projections are not allowed in subscriptions.
+
+### FROM
+
+```ebnf
+FROM table [ [AS] alias ] [ [INNER] JOIN table [ [AS] alias ] ON predicate ]
+```
+
+While you can only subscribe to rows from a single table,
+you may reference two tables in the `FROM` clause using a `JOIN`.
+A `JOIN` selects all combinations of rows from its input tables,
+and `ON` determines which combinations are considered.
+Note that if a `JOIN` is used,
+an index must be defined on the join column of both tables,
+so that it can be evaluated efficiently.
+
+### WHERE
+
+```ebnf
+predicate
+    = expr
+    | predicate AND predicate
+    ;
+
+expr
+    = literal
+    | column
+    | expr op expr
+    ;
+
+op
+    = '='
+    | '<'
+    | '>'
+    | '<' '='
+    | '>' '='
+    ;
+
+literal
+    = INTEGER
+    | STRING
+    | HEX
+    | TRUE
+    | FALSE
+    ;
+```
+
+While the `SELECT` clause determines the table,
+the `WHERE` clause determines the rows in the subscription.
+
+The subscription language supports conjunctions `AND`.
+It does not support disjunctions `OR`.
+It does not support `!=` comparisons.
+It does not support arithmetic expressions.
+Note that disjunctions can be expressed by subscribing to multiple queries.
+
+## Query and DML
+
+### Statements
+
+- [SELECT](#select)
+- [INSERT](#insert)
+- [DELETE](#delete)
+- [UPDATE](#update)
+- [SET](#set)
+- [SHOW](#show)
+
+### SELECT
+
+```ebnf
+SELECT [ DISTINCT ] projection FROM relation [ [ WHERE predicate ] [ ORDER BY order ] [ LIMIT limit ] ]
+```
+
+#### SELECT Clause
+
+```ebnf
+projection
+    = '*'
+    | table '.' '*'
+    | projExpr { ',' projExpr }
+    | aggrExpr { ',' aggrExpr }
+    ;
+
+projExpr
+    = column [ [ AS ] alias ]
+    ;
+
+aggrExpr
+    = COUNT '(' STAR ')' [AS] alias
+    | COUNT '(' DISTINCT column ')' [AS] alias
+    | SUM   '(' column ')' [AS] alias
+    ;
+```
+
+The `SELECT` clause determines the columns that are returned.
+In particular it supports both column and wildcard projections.
+It also supports sum and count aggregation.
+
+If `DISTINCT` is specified, only unique rows are returned.
+However each projected column must be of a type that allows comparison.
+
```

Comment:
Which types are not allowed? I think all of our `sats` types allow it, at least, all the `sql` ones that we support.

Reply:
This is the difference between SATS data types and the Spacetime SQL type system.

Take sum types for example. They have a binary representation, BSATN, with **a** total ordering, but that ordering is not unique, nor is it fixed. In fact the module is not even required to use it. The module could define its own ordering, or none at all.

So until orderings are tracked by the database, Spacetime SQL should be conservative in what it allows.